soroban-env-host/src/auth.rs

```diff
@@ -1,3 +1,197 @@
+//!      approved. Think of this as the case where your name is on a bank
+//!      account. Obviously if you ask for money out of that account, and you
+//!      provide a signature or PIN as an authentic credential, the withdrawal
+//!      is authorized. You are the designated owner or authority for it.
+//!
+//!   2. The action is labeled with _some set of rules_ that need to hold for
+//!      the action to be authorized, in which case we evaluate the rules in
+//!      context, whatever they may be. Think of this as the case where, say,
+//!      there are daily transfer limits on an account to minimize risks. Here
+//!      the rules have nothing to do with authority in the sense of a person,
+//!      merely the evaluation of a given situation.
+//!
+//! Some authorization scenarios will combine cases, and sometimes case 2 will
+//! delegate back to case 1 ("the rules" are "check for a signature from X").
+//! The examples above are illustrative of cases with the strictest difference,
+//! but that difference has crucial practical details:
+//!
+//!   - In case 1, the presence of a _valid signature_ (authentic credentials)
+//!     indicates authorization, and those are produced off-chain by private
+//!     keys. The authority is the corresponding public key (an XDR
+//!     `AccountID`).
+//!
+//!   - In case 2, the presence of _a call from a contract_ inside the current
+//!     transaction indicates authorization. We (the blockchain processor) know
+//!     which contracts ran in this transaction -- we ran them! -- and if any of
+//!     those contracts wants to decide, on the basis of its input, that some
+//!     action is authorized, all it has to do is tell us "I authorize this
+//!     action". No signatures are required (and indeed they would not make
+//!     sense, everything on a blockchain is public, there are no secret keys).
+//!
+//! This second case is subtle and worth repeating and emphasizing: in the
+//! simplest instance, contract A can "authorize" contract B by simply calling B
+//! itself, it doesn't much need to use an "auth module". A-calls-B is
+//! implicitly a statement that whatever logic exists in A, it decided it was
+//! correct to call B. This is such a basic fact that if B subsequently _asks_
+//! the auth module "did A authorize me?" in an A-called-B context, the auth
+//! module will just say "yes". We will see later there are other ways A might
+//! authorize B at some further distance than a direct call, but these are
+//! generalizations of the same principle: a contract _is_ at least partly an
+//! authorization policy, and if it says "yes" in a context where it has been
+//! designated an authority for the sake of authorization, then that "yes" is
+//! all the evidence we need.
+//!
+//! To sum up: addresses exist to generalize between these two cases: they allow
+//! a contract to define its authorization requirements in terms of _either_ a
```

Comment:
> //! a contract to define its authorization requirements in terms of _either_ a
> //! person (account), or a program (contract), or both.

The way I read this is that a contract can *choose* to define authorization requirements *only* for e.g. an account, which is not true. The requirement always abstracts the way the authorization is actually performed. Maybe I'm just not reading this correctly, but that still means that we probably need to reword this statement.